### PR TITLE
Add support for XEP-0363: HTTP File Upload.

### PR DESCRIPTION
**Pull request by grigoryfedorov**
Not sure about naming and Manager API. 
Hope it can be included in Smack 4.2.

### CODE REVIEW

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/packet/FileTooLargeError.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public CharSequence toXML() {
+        throw new IllegalStateException("Client should not return File Too Large error extension");
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/packet/FileTooLargeError.java]: This must be implemented for the debugger to work.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/packet/Slot.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public String getGetUrl() {
+        return getUrl;
+    }
+
+
+    @Override
+    protected IQChildElementXmlStringBuilder getIQChildElementBuilder(IQChildElementXmlStringBuilder xml) {
+        throw new IllegalStateException("Client should not return HTTP File Upload slot");
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/packet/Slot.java]: Dito

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/packet/Slot.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * @author Grigory Fedorov
+ * @see <a href="http://xmpp.org/extensions/xep-0363.html">XEP-0363: HTTP File Upload</a>
+ */
+public class Slot extends IQ {
+
+    public static final String ELEMENT = "slot";
+    public static final String NAMESPACE = SlotRequest.NAMESPACE;
+
+    private final String putUrl;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/packet/Slot.java]: Should be of type URL

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/packet/Slot.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    private final String getUrl;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/packet/Slot.java]: Should be of type URL

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 89ebe88e41d866a3d99c45320808ccca2feb530b
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/igniterealtime/Smack/wiki/Guidelines-for-Smack-Developers-and-Contributors

---
This message was auto-generated by https://gitcop.com

**Comment by grigoryfedorov** [comment]
Thanks for fast review! I will squash commits and add issue key when you approve pull request.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/element/SlotRequest.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    private final String filename;
+    private final long size;
+    private final String contentType;
+
+    public SlotRequest(String filename, long size, String contentType) {
+        super(ELEMENT, NAMESPACE);
+
+        this.filename = filename;
+        this.size = size;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/element/SlotRequest.java]: This should be sanitized, i.e. throw IllegalArgumentException when negative or zero.

**Comment by Flowdalic** [comment]
LGTM, it would be great if you could factor the typo fixes (thanks for that!) into an extra commit. Other then that, I think it's good to go.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            maxFileSize = Long.valueOf(values.get(0));
+        }
+
+        return uploadService;
+    }
+
+    /**
+     * Get upload service if available.
+     *
+     * Note that you must call {@link HttpFileUploadManager#discoverUploadService} first.
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: That's a little bit user unfriendly API wise. How about using a connection authenticated callback to automatically discover if a service exists and a boolean flag indicating if none was discovered. Then we also need a way to let the user manually specify the service.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        }
+
+        if (maxFileSize != null) {
+            if (fileSize > maxFileSize) {
+                throw new IllegalArgumentException("File size must not be greater than " + maxFileSize);
+            }
+        }
+
+        SlotRequest slotRequest = new SlotRequest(filename, fileSize, contentType);
+        slotRequest.setTo(uploadService);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: That assumes that `uploadService` was set by `discoverUploadService()`, which may not be true. Again the `uploadServiceDiscoveryed` flag should be used and a `SmackException` should be thrown is its false and if the user did not manually specify an upload service.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+     * @param contentType file content-type or null
+     * @return file upload Slot in case of success
+
+     * @throws IllegalArgumentException if fileSize is greater than {@link HttpFileUploadManager#getMaxFileSize()}
+     * @throws SmackException.NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPException.XMPPErrorException
+     * @throws SmackException.NoResponseException
+     */
+    public Slot requestSlot(String filename, long fileSize, String contentType) throws SmackException.NotConnectedException,
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: I'd also like to have a method `requestSlot(String, long, String, DomainBareJid)` where the user can specify the upload service to be used.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        XMPPConnectionRegistry.addConnectionCreationListener(new ConnectionCreationListener() {
+            @Override
+            public void connectionCreated(XMPPConnection connection) {
+                getInstanceFor(connection);
+            }
+        });
+    }
+
+    private static final Map<XMPPConnection, HttpFileUploadManager> INSTANCES = new WeakHashMap<>();
+    private DomainBareJid uploadService;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: That should be `defaultUploadService`

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java (hunk 6 above)
Comment (grigoryfedorov) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: If I add service discovery in authenticated callback, I have to catch and log exceptions. 
Upload service support can be unknown after error in discovering. 
So, flag should indicate success of discovery operation?

**Comment by Flowdalic** [comment]
Note to self: It would be great if the API would also do the Upload via HttpUrlConnection. Something like 

Url upload(File)

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java (hunk 6 above)
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: Simply have a member `DomainBareJid uploadService` and attempt to set it in the authenticated callback if not already set, and log all exceptions while doing so. I guess we don't even need a boolean flag.

**Comment by Flowdalic** [comment]
@grigoryfedorov Are you still up to this task? We could also go with the low level API sketch for HTTP upload that we have right now, and add the `Url upload(File file)` part later.

**Comment by grigoryfedorov** [comment]
@Flowdalic sorry for delay. 
I wrote `Url upload(File file)` using `HttpUrlConnection` just need some polishing.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 89ebe88e41d866a3d99c45320808ccca2feb530b
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/igniterealtime/Smack/wiki/Guidelines-for-Smack-Developers-and-Contributors

---
This message was auto-generated by https://gitcop.com

**Comment by grigoryfedorov** [comment]
@Flowdalic just to be sure that you noticed my last changes. I squashed all commits and made separate one for documentation corrections.

File: documentation/extensions/index.md [hunk 10]
```diff
@@ -21,41 +21,41 @@ Currently supported XEPs of smack-tcp
```
Comment (Flowdalic) [inline documentation/extensions/index.md]: Thanks for the fixes here, but could you put them into an extra commit?

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+     */
+    public URL uploadFile(File file) throws InterruptedException, XMPPException.XMPPErrorException,
+            SmackException, IOException {
+        return uploadFile(file, null);
+    }
+
+    /**
+     * Callback interface to get upload progress.
+     */
+    public interface UploadProgressListener {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: I'd rather have this in an extra file.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+     * Note that this is a synchronous call -- Smack must wait for the server response.
+     *
+     * @return true if upload service was discovered
+
+     * @throws XMPPException.XMPPErrorException
+     * @throws SmackException.NotConnectedException
+     * @throws InterruptedException
+     * @throws SmackException.NoResponseException
+     */
+    public boolean discoverUploadService() throws XMPPException.XMPPErrorException, SmackException.NotConnectedException,
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: Is there a reason to make this public?

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+            listener.onUploadProgress(0, fileSize);
+        }
+
+        BufferedInputStream inputStream = new BufferedInputStream(new FileInputStream(file));
+
+        byte[] buffer = new byte[4096];
+        int bytesRead;
+        while ((bytesRead = inputStream.read(buffer)) != -1) {
+            outputStream.write(buffer, 0, bytesRead);
+            outputStream.flush();
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: Wouldn't it be better to flush after the loop? A flush operation may degrade to a syscall, which is expensive.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+            bytesSend += bytesRead;
+
+            if (listener != null) {
+                listener.onUploadProgress(bytesSend, fileSize);
+            }
+
+        }
+
+        inputStream.close();
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java]: Aren't those two `close()` supposed to be within a finally block?

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/element/FileTooLargeError.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public CharSequence toXML() {
+        XmlStringBuilder xml = new XmlStringBuilder(this);
+        xml.rightAngleBracket();
+        xml.element("max-file-size", String.valueOf(maxFileSize));
+        xml.closeElement(ELEMENT);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/element/FileTooLargeError.java]: Should be `xml.closeElement(this)`

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/element/FileTooLargeError.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        return ELEMENT;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public CharSequence toXML() {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/element/FileTooLargeError.java]: Return type should be `XmlStringBuilder`

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/element/SlotRequest.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+        return contentType;
+    }
+
+    @Override
+    protected IQChildElementXmlStringBuilder getIQChildElementBuilder(IQChildElementXmlStringBuilder xml) {
+        xml.rightAngleBracket();
+        xml.element("filename", filename);
+        xml.element("size", String.valueOf(size));
+        if (contentType != null) {
+            xml.element("content-type", contentType);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/element/SlotRequest.java]: Use `xml.optElement(String, String)`

**Comment by Flowdalic** [comment]
Squashed and cherry picked you work with https://github.com/igniterealtime/Smack/commit/72d4c8b611299e50922d23eab3f864cb11dceadc and fixed a few things with https://github.com/igniterealtime/Smack/commit/09b6608a3af963e054319e3e9d67437400e99e5b

Thanks for your contribution.

**Comment by grigoryfedorov** [comment]
@Flowdalic thanks for improvements and merge (special thanks for 0.2 support).
And sorry for not updating in time.

**Pull request closed without merging**